Improve mouse-based load inhibition at player loading screen

UselessToucan opened the pull request:
We probably do not want to move from PlayerLoader to Player if the user pressed mouse button down on one of the sliderbars and did not release it yet.

I can't provide the video right now, but here are the steps to reproduce:
1. Perform MouseDown on one of the VisualSettings overlay sliderbars.
2. Move the cursor outside of the overlay while holding mouse button pressed.

peppy reviewed (changes requested):
Still not a huge fan of exposing IsMouseDown.

smoogipoo commented:
How about just using `Drawable.IsDragged` on the sliderbars?

UselessToucan commented:
@smoogipoo I tried to use `Drawable.IsDragged`. It does not solve the problem since it becomes false earlier than I release the mouse button.

peppy commented:
According to its implementation, it shouldn't.

peppy commented:
Two potential better solutions:

In `PlayerLoader.cs`

```csharp
var mouseDownPos = GetContainingInputManager().CurrentState.Mouse.NativeState.PositionMouseDown;
if (player.LoadState != LoadState.Ready || visualSettings.IsHovered || (mouseDownPos != null && visualSettings.ReceiveMouseInputAt(mouseDownPos.Value)))
{
    Schedule(pushWhenLoaded);
    return;
}
```

In `VisualSettings.cs`
```csharp
private bool mouseDown;

public override bool ReceiveMouseInputAt(Vector2 screenSpacePos) => mouseDown || base.ReceiveMouseInputAt(screenSpacePos);

protected override bool OnMouseDown(InputState state, MouseDownEventArgs args)
{
    mouseDown = true;
    return base.OnMouseDown(state, args);
}

protected override bool OnMouseUp(InputState state, MouseUpEventArgs args)
{
    mouseDown = false;
    return base.OnMouseUp(state, args);
}
```

Both result in very localised implementations, which handle more than just sliders at the same time. Not sure which I prefer.

UselessToucan commented:
I prefer the first one. In my opinion, it is easier.

peppy commented:
@smoogipoo would like a second opinion on this. it's likely going to pop up in more places going forward if anything (volume overlay being one such example).

smoogipoo commented:
At the very least I believe these should be inverted. I.e. instead of `visualSettings.ReceiveMouseInputAt(...)` it should use `PlayerLoader.ReceiveMouseInputAt`. Same thing with `IsHovered`.

Furthermore, why does `IsDragged` not work? Sounds like an issue...

peppy commented:
I checked on `IsDragged`: you can potentially leave the visual settings area before the drag lenience has been hit, at which point there's no dragged state activated.

Inverting could have unwanted side effects, like hovering the toolbar or chat or something. Not sure if we want that.

smoogipoo commented:
The idea being that those places should not handle OnHover if they don't want player's loading to stop.

I mean there really isn't any other way - either you have the player check for each component that it needs, or you have some "hey I'm currently handling input" state be passed down to the `PlayerLoader`.

I believe the chat is an overlaycontainer, yes? In which case it probably already does block passthrough hover and I think that makes more sense than pushing a player while chat is open?

peppy commented:
It's definitely worth a try to see how it feels.

peppy committed:
Block player enter when a drag initiates from an overlaying container

peppy committed:
Add debouncing to player loading 

Allows the mouse to temporarily exit and re-enter overlay elements without triggering a load

peppy commented:
Reworked to add debounce and work in a more general fashion. @smoogipoo requesting review as this is now mostly my code ;x

smoogipoo commented on osu.Game/Screens/Play/PlayerLoader.cs:
```diff
@@ -116,29 +118,52 @@ protected override void LogoArriving(OsuLogo logo, bool resuming)
         private void pushWhenLoaded()
         {
-            if (player.LoadState != LoadState.Ready || visualSettings.IsHovered)
+            Schedule(pushWhenLoaded);
```
This scheduling is still going on during gameplay

smoogipoo commented on osu.Game/Screens/Play/PlayerLoader.cs:
```diff
@@ -116,29 +118,52 @@ protected override void LogoArriving(OsuLogo logo, bool resuming)
-            contentOut();
-
-            this.Delay(250).Schedule(() =>
+            if (pushDebounce == null) pushDebounce = Scheduler.AddDelayed(() =>
```
Put this on a separate line please.

peppy committed:
Merge remote-tracking branch 'upstream/master' into DoNotMoveToPlayerWhileDragging

peppy committed:
Apply formatting changes

peppy committed:
Avoid scheduling during non-current screen

peppy committed:
Merge branch 'master' into DoNotMoveToPlayerWhileDragging
